sdk/core/Azure.Core/src/Shared/TaskExtensions.cs

```diff
@@ -133,5 +157,80 @@ private static void VerifyTaskCompleted(bool isCompleted)
             public void Dispose() => _asyncEnumerator.DisposeAsync().EnsureCompleted();
 #pragma warning restore AZC0107 // Do not call public asynchronous method in synchronous scope.
         }
+
+        public readonly struct WithCancellationAwaitable<T>
```

Comment:
I'm worried about this implementation being a bit complicated.

Can we go with something simpler like:
            
``` C#
var tcs = new TaskCompletionSource<T>();
_registration = cancellationToken.Register(() => tcs.TrySetCanceled(cancellationToken));

await Task.WhenAny(tcs.Task, task)
```

??

Reply:
- With a switch to `ValueTask`, `Task.WhenAny` can't be used directly and will require `AsTask()` to be called
- `Task.WhenAny(tcs.Task, task).Unwrap().ConfigureAwait(false)` is much slower (because of all the task continuation overhead) and produces a lot of garbage in memory. Awaiter in the worst case requires only two objects - continuation wrapper and delegate.

Reply:
Currently I'm worried about future supportability of this code.

It doesn't like any of the current uses is perf sensitive.

Reply:
I don't know how often `BearerTokenAuthenticationPolicy.ProcessAsync` can be called, however there are at least 2 more places in `Azure.Identity` where this method will be useful, so I'd like to make it publicly available. Which means we won't know where and how often it will be used.